src/System.Net.Security/src/System/Net/_SSPIWrapper.cs

```diff
@@ -0,0 +1,873 @@
+    internal class StreamSizes
+    {
+        public int header;
+        public int trailer;
+        public int maximumMessage;
+        public int buffersCount;
+        public int blockSize;
+
+        internal unsafe StreamSizes(byte[] memory)
+        {
+            fixed (void* voidPtr = memory)
+            {
+                IntPtr unmanagedAddress = new IntPtr(voidPtr);
+                try
+                {
+                    header = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress));
+                    trailer = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress, 4));
+                    maximumMessage = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress, 8));
+                    buffersCount = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress, 12));
+                    blockSize = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress, 16));
+                }
+                catch (OverflowException)
+                {
+                    GlobalLog.Assert(false, "StreamSizes::.ctor", "Negative size.");
+                    throw;
+                }
+            }
+        }
+        public static readonly int SizeOf = Marshal.SizeOf<StreamSizes>();
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal class SecSizes
+    {
+        public readonly int MaxToken;
+        public readonly int MaxSignature;
+        public readonly int BlockSize;
+        public readonly int SecurityTrailer;
+
+        internal unsafe SecSizes(byte[] memory)
+        {
+            fixed (void* voidPtr = memory)
+            {
+                IntPtr unmanagedAddress = new IntPtr(voidPtr);
+                try
+                {
+                    MaxToken = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress));
+                    MaxSignature = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress, 4));
+                    BlockSize = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress, 8));
+                    SecurityTrailer = (int)checked((uint)Marshal.ReadInt32(unmanagedAddress, 12));
+                }
+                catch (OverflowException)
+                {
+                    GlobalLog.Assert(false, "SecSizes::.ctor", "Negative size.");
+                    throw;
+                }
+            }
+        }
+        public static readonly int SizeOf = Marshal.SizeOf<SecSizes>();
+    }
+
+    // TODO (Issue #3114): Move to Interop. 
+    // From Schannel.h
+    [StructLayout(LayoutKind.Sequential)]
+    internal class SslConnectionInfo
+    {
+        public readonly int Protocol;
+        public readonly int DataCipherAlg;
+        public readonly int DataKeySize;
+        public readonly int DataHashAlg;
+        public readonly int DataHashKeySize;
+        public readonly int KeyExchangeAlg;
+        public readonly int KeyExchKeySize;
+
+        internal unsafe SslConnectionInfo(byte[] nativeBuffer)
+        {
+            fixed (void* voidPtr = nativeBuffer)
```

Comment:
Added TODO. The code isn't safe as written: it should cast to a struct instead (`Marshal.PtrToStructure`).